examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3721 @@
+            host.invoke_contract(
+                &address,
+                &input.params.data,
+                function.as_entrypoint_name(),
+                unwrapped_amount,
+            )?;
+        }
+    }
+
+    // Log the burn event.
+    let parameter_bytes = to_bytes(&BurnEventParams {
+        amount:  input.params.amount,
+        address: input.params.owner,
+    });
+
+    host.invoke_contract_raw(
+        &proxy_address,
+        Parameter(&parameter_bytes),
+        EntrypointName::new_unchecked("logBurnEvent"),
+        Amount::zero(),
+    )?;
+
+    Ok(())
+}
+
+// Contract functions required by CIS2
+
+pub type TransferParameter = TransferParams<ContractTokenId, ContractTokenAmount>;
+
+/// Execute a list of token transfers, in the order of the list.
+///
+/// Logs a `Transfer` event and invokes a receive hook function for every
+/// transfer in the list.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - Any of the transfers fail to be executed, which could be if:
+///     - The `token_id` does not exist.
+///     - The sender is not the owner of the token, or an operator for this
+///       specific `token_id` and `from` address.
+///     - The token is not owned by the `from`.
+/// - Fails to log event.
+/// - Any of the receive hook function calls rejects.
+#[receive(contract = "CIS2-wCCD", name = "transfer", mutable)]
+fn contract_transfer<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    let (proxy_address, state_address) = if let ProtocolAddressesImplementation::Initialized {
+        proxy_address,
+        state_address,
+    } = host.state().protocol_addresses
+    {
+        (proxy_address, state_address)
+    } else {
+        panic!("{:?}", concordium_cis2::Cis2Error::Custom(CustomContractError::UnInitialized));
+    };
+
+    // Can be only called through the fallback function on the proxy.
+    only_proxy(proxy_address, ctx.sender())?;
+
+    // Check that contract is not paused.
+    when_not_paused(host)?;
+
+    // Parse the parameter.
+    let input: TransferParameterWithSender = ctx.parameter_cursor().get()?;
+
+    let TransferParams(transfers) = input.params;
+
+    for Transfer {
+        token_id,
+        amount,
+        from,
+        to,
+        data,
+    } in transfers
+    {
+        // Get the sender who invoked this contract function.
+        let sender = input.sender;
+
+        // Authenticate the sender for this transfer
+        ensure!(
+            from == sender
+                || host
+                    .state()
+                    .is_operator(&sender, &from, host)
+                    .map_or(false, |is_operator| is_operator),
+            ContractError::Unauthorized
+        );
+        let to_address = to.address();
+        // Update the contract state
+
+        if amount == 0u64.into() {
+            return Ok(());
+        }
```

Comment:
1. This is returning the entire transfer function call, meaning the rest of the loop is never executed after a zero amount transfer.

2. According to the CIS-2:  _A transfer of a token amount zero MUST be executed as a normal transfer_.
So using `continue` here is right either. It is fine to skip calling the state contract in this case, but it will still need to log a transfer event and call the hook.

Reply:
Transfer events are now logged even if the amount transferred is 0.